File: medium/typescript/49-group-anagrams.ts
function groupAnagrams(strs: string[]): string[][] {
  const anagramGroups = new Map();

  for (const str of strs) {
    const count = new Array(26).fill(0);

    for (const char of str) {
      const index = char.charCodeAt(0) - "a".charCodeAt(0);
      count[index]++;
    }

    const key = count.join("#");

    if (!anagramGroups.has(key)) {
      anagramGroups.set(key, []);
    }
    anagramGroups.get(key).push(str);
  }

  return Array.from(anagramGroups.values());
}

function groupAnagrams1(strs: string[]): string[][] {
  const anagramGroups = new Map<string, string[]>();

  for (const str of strs) {
    const sortedStr = str.split("").sort().join("");

    if (anagramGroups.has(sortedStr)) {
      anagramGroups.get(sortedStr)?.push(str);
    } else {
      anagramGroups.set(sortedStr, [str]);
    }
  }

  return Array.from(anagramGroups.values());
}

File: medium/python/49-group-anagrams.py
class Solution:
    def group_anagrams(self, strs: list[str]) -> list[list[str]]:
        group_anagrams: dict[str, list[str]] = {}

        for word in strs:
            alphabet: list[int] = [0] * 26

            for char in word:
                index = ord(char) - ord("a")
                alphabet[index] += 1

            key = "#".join(map(str, alphabet))

            if key not in group_anagrams:
                group_anagrams[key] = [word]
            else:
                group_anagrams.get(key).append(word)

        return list(group_anagrams.values())

    def groupAnagrams1(self, strs: list[str]) -> list[list[str]]:
        group_anagrams: dict[str, list[str]] = {}

        for word in strs:
            sorted_list = sorted(word)
            sorted_word = "".join(sorted_list)

            if sorted_word in group_anagrams:
                group_anagrams.get(sorted_word).append(word)
            else:
                group_anagrams[sorted_word] = [word]

        return list(group_anagrams.values())
